libraries/chain/block_log.cpp

```diff
@@ -469,20 +478,243 @@ namespace eosio { namespace chain {
+      using block_num_t = uint32_t;
+
+      struct mapped_type {
+         block_num_t             last_block_num;
+         boost::filesystem::path filename_base;
+      };
+      using collection_t              = boost::container::flat_map<block_num_t, mapped_type>;
+      using size_type                 = collection_t::size_type;
+      static constexpr size_type npos = std::numeric_limits<size_type>::max();
+
+      boost::filesystem::path    archive_dir;
+      size_type                  max_retained_files = 10;
+      collection_t               collection;
+      size_type                  active_index = npos;
+      block_log_data             log_data;
+      block_log_index            log_index;
+      chain_id_type              chain_id;
+
+      bool empty() const { return collection.empty();  }
+
+      void open(fc::path block_dir) {
+         for_each_file_in_dir_matches(block_dir, R"(blocks-\d+-\d+\.log)", [this](boost::filesystem::path path) {
+            auto log_path               = path;
+            auto index_path             = path.replace_extension("index");
+            auto path_without_extension = log_path.stem().string();
+
+            block_log_data log(log_path);
+
+            if (chain_id.empty()) {
+               chain_id = log.chain_id();
+            } else {
+               EOS_ASSERT(chain_id == log.chain_id(), block_log_exception, "block log file ${path} has a different chain id",
+                           ("path", log_path.generic_string()));
+            }
+
+            // check if index file matches the log file
+            if (!index_matches_data(index_path, log))
+               block_log::construct_index(log_path, index_path);
+
+            auto [itr, _]      = collection.emplace(log.first_block_num(),
+                                                    mapped_type{ log.last_block_num(), path_without_extension });
+            this->active_index = collection.index_of(itr);
+         });
+      }  
+
+      bool index_matches_data(const boost::filesystem::path& index_path, const block_log_data& log) const {
+         if (boost::filesystem::exists(index_path) &&
+             boost::filesystem::file_size(index_path) / sizeof(uint64_t) != log.num_blocks()) {
+            // make sure the last 8 bytes of index and log matches
+
+            fc::cfile index_file;
+            index_file.set_file_path(index_path);
+            index_file.open("r");
+            index_file.seek_end(-sizeof(uint64_t));
+            uint64_t pos;
+            index_file.read(reinterpret_cast<char*>(&pos), sizeof(pos));
+            return pos == log.last_block_position();
+         }
+         return false;
+      }
+
+      bool set_active_item(uint32_t block_num) {
+         try {
+            if (active_index != npos) {
+               auto active_item = collection.nth(active_index);
+               if (active_item->first <= block_num && block_num <= active_item->second.last_block_num) {
```

Comment:
My thought is we should care.

Reply:
My thought is that we shouldn't care. If the user has block log files with overlapping ranges and the blocks in the overlapping ranges have different contents, that means at least one the block log has been manually tempered by user deliberately. There's no good way to prevent this from this kind of attack even block logs doesn't have overlapping ranges; unless we want to validate the integrity of all block log files during loading. I am fine with adding an option for this kind of validation. However, saying that the situation only happen when the block logs files have overlapping ranges doesn't make a lot sense to me.